FirebaseMLModelDownloader/Sources/ModelDownloader.swift

```diff
@@ -136,3 +155,71 @@ public class ModelDownloader {
     completion(.failure(.notFound))
   }
 }
+
+extension ModelDownloader {
+  /// Return local model info only if the model info is available and the corresponding model file is already on device.
+  private func getLocalModelInfo(modelName: String) -> LocalModelInfo? {
+    guard let localModelInfo = LocalModelInfo(
+      fromDefaults: userDefaults,
+      name: modelName,
+      appName: appName
+    ),
+      let localPath = URL(string: localModelInfo.path),
+      ModelFileManager.isFileReachable(at: localPath) else {
+      // TODO: Delete local model info in user defaults
+      return nil
+    }
+    return localModelInfo
+  }
+
+  /// Get model saved on device if available.
+  private func getLocalModel(modelName: String) -> CustomModel? {
+    guard let localModelInfo = getLocalModelInfo(modelName: modelName) else { return nil }
+    let model = CustomModel(localModelInfo: localModelInfo)
+    return model
+  }
+
+  /// Download and get model from server, unless the latest model is already available on device.
+  private func getRemoteModel(modelName: String,
+                              progressHandler: ((Float) -> Void)? = nil,
+                              completion: @escaping (Result<CustomModel, DownloadError>) -> Void) {
+    let localModelInfo = getLocalModelInfo(modelName: modelName)
+    let modelInfoRetriever = ModelInfoRetriever(
+      modelName: modelName,
+      options: options,
+      installations: installations,
+      appName: appName,
+      localModelInfo: localModelInfo
+    )
+
+    modelInfoRetriever.downloadModelInfo { result in
+      switch result {
+      case let .success(remoteModelInfo):
+        /// New model info was downloaded from server.
+        if let remoteModelInfo = remoteModelInfo {
+          let downloadTask = ModelDownloadTask(
+            remoteModelInfo: remoteModelInfo,
+            appName: self.appName,
+            defaults: self.userDefaults,
+            progressHandler: progressHandler,
+            completion: completion
+          )
+          downloadTask.resumeModelDownload()
+        } else {
+          guard let localModel = self.getLocalModel(modelName: modelName) else {
+            /// This can only happen if local model info was suddenly wiped out in the middle of model info request and server response.
+            completion(
+              .failure(
+                .internalError(description: "Model unavailable due to deleted local model info.")
+              )
+            )
+            return
+          }
+          completion(.success(localModel))
+        }
+      case let .failure(downloadError):
+        completion(.failure(downloadError))
+      }
```

Comment:
Can we wrap these completion handlers to ensure they're on the main thread?

Reply:
Done!